File: Tema1/V2/Server/src/logic/TCPServer.py
import socket

from threading import Thread, Lock
from time import time

from .BenchmarkData import BenchmarkData


class TCPServer:
    def __init__(self, host, port, benchmark_file_path, chunk_size, ack):
        self.__host = host
        self.__port = port
        self.__running = False
        self.__running_lock = Lock()
        self.__benchmark_file_path = benchmark_file_path
        self.__chunk_size = chunk_size
        self.__ack = ack

    def run(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            sock.bind((self.__host, self.__port))
            sock.listen()

            print(f"Listening on {self.__host}:{self.__port}")

            self.__set_running(True)

            while self.__get_running():
                conn, addr = sock.accept()
                handle_client_thread = Thread(target=self.__handle_client, args=(conn, addr,))
                handle_client_thread.start()

    def stop(self):
        self.__set_running(False)

    def __handle_client(self, conn, addr):
        client_ip = addr[0]

        print(f"{client_ip} connected")

        bytes_count = 0

        with open(self.__benchmark_file_path, "wb") as benchmark_file:
            while self.__get_running() and conn:
                chunk = conn.recv(self.__chunk_size)

                if len(chunk) == 0:
                    print(f"{client_ip} sent {bytes_count} bytes")
                    conn.close()

                    category = 'tcp_server' if self.__ack else 'tcp_stream_server'
                    BenchmarkData.add_data(client_ip, category, time(), bytes_count)

                    return

                benchmark_file.write(chunk)

                bytes_count += len(chunk)

                if self.__ack:
                    self.__acknowledge(conn)

    def __acknowledge(self, conn):
        try:
            ack_bytes = bytearray()
            ack_bytes.append(1)
            conn.send(ack_bytes)
        except:
            pass

    def __set_running(self, running):
        self.__running_lock.acquire()
        self.__running = running
        self.__running_lock.release()

    def __get_running(self):
        self.__running_lock.acquire()
        running = self.__running
        self.__running_lock.release()
        return running

File: Tema1/Server/src/network/base/TCPServer/TCPServer.cpp
#include "TCPServer.h"


TCPServer::TCPServer(int port) {
    this->port = port;
    this->running = false;
    this->runningMutex = new std::mutex();
}

TCPServer::~TCPServer() {
    delete this->runningMutex;
}

void TCPServer::run() {
    if (getRunning()) {
        throw std::logic_error("Server already running");
    }
    setRunning(true);

    int sockDesc = socket(AF_INET, SOCK_STREAM, 0);
    if (sockDesc < 0) {
        throw std::runtime_error("Failed to open socket");
    }

    struct sockaddr_in sockAddr = {};
    sockAddr.sin_family = AF_INET;
    sockAddr.sin_addr.s_addr = INADDR_ANY;
    sockAddr.sin_port = htons(port);
    int bindRes = bind(sockDesc, (struct sockaddr *)&sockAddr, sizeof(sockAddr));
    if (bindRes < 0) {
        throw std::runtime_error("Failed to bind socket");
    }

    std::thread serverDaemonThread(&TCPServer::serverDaemon, this, sockDesc);
    serverDaemonThread.detach();
}

void TCPServer::stop() {
    if (!getRunning()) {
        throw std::logic_error("Server not running");
    }
    setRunning(false);
}

void TCPServer::serverDaemon(int sockDesc) {
    std::cout<<"Listening for connections\r\n";
    listen(sockDesc, SOMAXCONN);

    while (getRunning()) {
        struct sockaddr_in clientSockAddr;
        socklen_t clientSockAddrSize = sizeof(clientSockAddr);
        int clientSockDesc = accept(sockDesc, (struct sockaddr *)&clientSockAddr, &clientSockAddrSize);
        if (clientSockDesc < 0) {
            std::cerr<<"Failed to accept client connection\r\n";
            continue;
        }
        char * clientIP = inet_ntoa(clientSockAddr.sin_addr);
        std::cout<<"Accepted client connection from "<<clientIP<<"\r\n";

        std::thread handleClientThread(&TCPServer::handleClientEncaps, this, clientSockDesc, clientIP);
        handleClientThread.detach();
    }

    close(sockDesc);
}

void TCPServer::handleClientEncaps(int clientSockDesc, char * clientIP) {
    handleClient(clientSockDesc, clientIP);

    close(clientSockDesc);
}

bool TCPServer::getRunning() {
    runningMutex->lock();
    bool runningVal = running;
    runningMutex->unlock();

    return runningVal;
}

void TCPServer::setRunning(bool running) {
    runningMutex->lock();
    this->running = running;
    runningMutex->unlock();
}